src/index.ts

```diff
@@ -140,70 +159,103 @@ export class Cinnamon {
+            } catch (error) {
+                throw new AdgoError(error);
+            }
 
-            json = await response.json();
+            if (json.errors) {
+                if (
+                    json.errors.some(
+                        (error: APIError) =>
+                            get(error, 'extensions.code') ===
+                            codes.TOKEN_EXPIRED,
+                    ) &&
+                    !token
+                ) {
+                    if (!this.refreshTokenRequest) {
+                        this.refreshTokenRequest = this.refreshLogin({
+                            refreshToken: this.refreshToken,
+                        }).then(() => null);
+                    }
+
+                    this.refreshTokenRequest = await this.refreshTokenRequest;
+                    return this.api({ query, variables, headers, token });
+                }
+                throw new CinnamonError(
+                    json.errors.map((error: Error) => error.message).join('\n'),
+                    json,
+                );
+            }
+            if (!json.data) {
+                throw new CinnamonError(
+                    `Invalid server response: ${JSON.stringify(json)}`,
+                    json,
+                );
+            }
+            return json;
         } catch (error) {
-            throw new AdgoError(error);
-        }
+            if (retryCount >= this.config.maxRetry) {
+                throw error;
+            }
 
-        if (json.errors) {
             if (
-                json.errors.some(
-                    (error: APIError) =>
-                        get(error, 'extensions.code') === codes.TOKEN_EXPIRED,
-                ) &&
-                !token
+                error?.raw?.errors?.some((error: APIError) => {
+                    const code = get(error, 'extensions.code');
+                    return code && !isRetryableCode(code);
+                })
             ) {
-                if (!this.refreshTokenRequest) {
-                    this.refreshTokenRequest = this.refreshLogin({
-                        refreshToken: this.refreshToken,
-                    }).then(() => null);
-                }
+                throw error;
+            }
 
-                this.refreshTokenRequest = await this.refreshTokenRequest;
-                return this.api({ query, variables, headers, token });
+            if (
+                this.config.retryHook &&
+                !(await this.config.retryHook(error, retryCount))
+            ) {
+                throw error;
             }
-            throw new CinnamonError(
-                json.errors.map((error: Error) => error.message).join('\n'),
-                json,
-            );
-        }
-        if (!json.data) {
-            throw new CinnamonError(
-                `Invalid server response: ${JSON.stringify(json)}`,
-                json,
+
+            await sleep(this.config.retrySleepTime * retryCount);
```

Comment:
For the first retry, it will always sleep `0`, because `retrySleepTime * 0`.

Reply:
yes, thats by design